Data/Map/Lazy.hs

```diff
@@ -14,19 +14,63 @@
 -- Maintainer  :  libraries@haskell.org
 -- Portability :  portable
 --
--- An efficient implementation of ordered maps from keys to values
--- (dictionaries).
 --
--- API of this module is strict in the keys, but lazy in the values.
--- If you need value-strict maps, use "Data.Map.Strict" instead.
--- The 'Map' type itself is shared between the lazy and strict modules,
--- meaning that the same 'Map' value can be passed to functions in
--- both modules (although that is rarely needed).
+-- = Lazy Maps (aka dictionaries)
 --
--- These modules are intended to be imported qualified, to avoid name
--- clashes with Prelude functions, e.g.
+-- The @'Map' k v@ type represents an /ordered/ map indexed by keys of type @k@
+-- containing /lazy/ values of type @v@.
 --
--- >  import qualified Data.Map.Lazy as Map
+-- When deciding if this is the correct data structure to use, consider:
+--
+-- * The API of this module is strict in the keys but lazy in the values. If you
+-- don't need /value-lazy/ maps, use "Data.Map.Strict" instead.
+--
+-- * If your key type is @Int@, use "Data.IntMap.Lazy" instead.
+--
+-- * If you don't care about ordering, use @Data.HashMap.Lazy@ from the
+-- <https://hackage.haskell.org/package/unordered-containers unordered-containers>
+-- package instead.
```

Comment:
This is often right, but not always. Size calculations for `HashMap` are very slow (although someone wants to try to fix that, I'm not sure it's going to happen). `Map` unions also have much tighter performance bounds than `HashMap` ones do. I would just say that people should consider using `unordered-containers` instead.